New catalog-backend-module-puppetdb plugin

tdabasinskas opened the pull request:
Hello,

I'm adding a new Backend Module plugin `catalog-backend-module-puppetdb`, which allows ingesting nodes from PuppetDB into Backstage as Resource Entities. This is useful for the companies that use Puppet as their configuration management solution and want to see their infrastructure assets (i.e., servers) in Backstage.

The default implementation imports the nodes, using the certificate name as the entity name and using the node IP address as the Entity description. Custom transformers allow users to define their own logic, using all available Puppet facts for constructing the Resource information.

The plugin also, by default, adds `puppet.com/certname` annotation to all resources, so we could identify Resources coming from PuppetDB in Backstage easier. This potentially allows developing a frontend plugin, which would add `Puppet` section for such resources, allowing to see information about the last Puppet runs.

---

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [X] Added or updated documentation
- [X] Tests for new functionality and regression tests for bug fixes
- [X] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))
---

tdabasinskas commented:
I see E2E tests are failing, but I don't think it has anything to do with my changes 😕

taras commented on plugins/catalog-backend-module-puppetdb/src/providers/PuppetDbEntityProvider.ts:
```diff
@@ -0,0 +1,236 @@
+  /** {@inheritdoc @backstage/plugin-catalog-backend#EntityProvider.connect} */
+  async connect(connection: EntityProviderConnection): Promise<void> {
+    this.connection = connection;
+    await this.scheduleFn();
```
You don't need to setup the task after connection is created. The scheduler won't run any tasks before entry providers are setup. You can create the task in fromConfig. The structure of your plug-in becomes non-conditional, but this will work too.

tdabasinskas replied:
Thanks!

As for the scheduler task, I did this for consistency, as it seems [all other catalog backend plugins are doing the same](https://github.com/search?q=repo%3Abackstage%2Fbackstage%20await%20this.scheduleFn()%3B&type=code).

Should I move `await this.scheduleFn();` to `fromConfig` instead?

taras replied:
No, it's fine. That's the correct way

taras reviewed (approved):
This is good looking open source: good documentation, tests and clean code. 👍

freben commented on plugins/catalog-backend-module-puppetdb/src/providers/PuppetDbEntityProvider.ts:
```diff
@@ -0,0 +1,236 @@
+ */
+function withLocations(host: string, entity: Entity): Entity {
+  const location = new URL(host);
+  location.pathname = `${ENDPOINT_NODES}/${entity.metadata?.name}`;
```
Can the puppetdb ever be configured on a subpath? Could you have it on `https://infra.mycompany.net/puppetdb`?

Maybe you should append the path to the base instead (ensuring that they get joined with a single slash, whether the app-config accidentally had an extra trailing slash in it or not; maybe the config reader can strip out those already at the source)

tdabasinskas replied:
We already trim the extra trailing slash as per https://github.com/backstage/backstage/pull/16184/files#diff-0ce4d0baa04b46c5b299bdd09d351d94c160e92bfa54a3a76c1e3072cc900180R86

freben reviewed (commented):
Excellent work! A number of comments but most are small, happy to get this in soon!

freben commented:
Note also that a rebase on new master is needed (rebase is better than reverse merge!), since we've fixed those pesky unrelated end-to-end test failures that weren't your fault

freben commented:
@tdabasinskas Gentle ping on this one!

tdabasinskas commented on plugins/catalog-backend-module-puppetdb/src/puppet/transformers.ts:
```diff
@@ -0,0 +1,64 @@
+      },
+      namespace: DEFAULT_NAMESPACE,
+      description: node.facts?.data
+        ?.find(e => e.name === 'ipaddress')
```
I was initially thinking about adding it as an annotation, but since `Description` is shown in entities list by default, I thought it might be a good idea to fill it with IP 🤷

tdabasinskas commented on plugins/catalog-backend-module-puppetdb/src/puppet/transformers.ts:
```diff
@@ -0,0 +1,64 @@
+      description: node.facts?.data
+        ?.find(e => e.name === 'ipaddress')
+        ?.value?.toString(),
+      tags: kernel ? [kernel.toString().toLowerCase()] : [],
```
Yes, it should contain only values of `linux`, `windows`, `aix`, `darwin`, `sunos`.

Based on examples in other Backstage places, like using programming language names as tags, I though using OS as a tag by default might make sense.

Both for this, and for IP Address comments above – I don't think there be many cases where the default transformer will be used. Most of the users will want to extract more data from Puppet Facts and control what goes to annotations.

tdabasinskas commented on plugins/catalog-backend-module-puppetdb/src/providers/PuppetDbEntityProviderConfig.ts:
```diff
@@ -0,0 +1,99 @@
+  /**
+   * (Required) The host of PuppetDB API instance.
+   */
+  host: string;
```
Yeah, I was considering this.

What if we rename it to `baseUrl`? I'm not sure if `puppetdb` prefix is necessary, as this setting is already under `puppetDb` provider, so it's clear it's for Puppet DB.

tdabasinskas commented:
> @tdabasinskas Gentle ping on this one!

Sorry for the delay ant thanks for review!

I've addressed most of the comments. Still need to look into the ones related to tests – will try to do that in upcoming days.

It would be also great if we could decide how to proceed with a few other items (like the one with `host` configuration parameter) I left comments on – I could address them as well if needed.

tdabasinskas commented on plugins/catalog-backend-module-puppetdb/src/providers/PuppetDbEntityProvider.test.ts:
```diff
@@ -0,0 +1,217 @@
+
+  describe('where there are no nodes', () => {
+    beforeEach(() => {
+      // @ts-ignore
```
Agreed regarding `// @ts-ignore`. However, looking into more examples in this repository, I see we are using `jest.spyOn` in a lot of places, which does the job and does not require `jest.spyOn`.

To avoid major refactoring (converting this into a class), I've switched to `jest.spyOn`. Actually, this even allowed catching a bug in the tests (the mocked entity structure was actually not correct) 🙂. Hope it's OK with you.

tdabasinskas commented:
Hey @freben,

I've finalized addressing the review. Let me know if anything else is needed.

taras commented:
@tdabasinskas thank you for the contribution! @freben thank you for the reviews!
